[WebGPU] Only the first item in PassDescriptor.timestampWrites actually works

litherum opened the pull request:
#### d91e1e789dfe516c82b6b2e03dc9603b19a908d4
<pre>
[WebGPU] Only the first item in PassDescriptor.timestampWrites actually works
<a href="https://bugs.webkit.org/show_bug.cgi?id=251779">https://bugs.webkit.org/show_bug.cgi?id=251779</a>
&lt;rdar://problem/105072144&gt;

Reviewed by Tadeu Zagallo and Dean Jackson.

rdar://91371495 is about how we can&apos;t just naively transform descriptor.timestampWrites into computePassDescriptor.sampleBufferAttachments.
Instead, we can resolve all the information to a dummy counter sample buffer, and then internally remember that the data
is in a different place than where it&apos;s supposed to be. Later, when we resolve the data, we can resolve it from our dummy
buffer instead of from where it&apos;s supposed to be.

When rdar://91371495 is fixed, we can delete this indirection, and put the data directly where it&apos;s supposed to go.

This patch causes query-sets.html to work.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateComputePassDescriptor const):
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::resolveQuerySet):
* Source/WebGPU/WebGPU/QuerySet.h:
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::QuerySet::QuerySet):
(WebGPU::QuerySet::destroy):
(WebGPU::QuerySet::setOverride):
(WebGPU::QuerySet::resolveTimestamps const):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::endPass): This is necessary because pass encoders internally retain query sets. If you allocate too many
query sets, trying to create more will fail. So, make sure we don&apos;t leave too many query sets hanging around.
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js:

Canonical link: <a href="https://commits.webkit.org/259925@main">https://commits.webkit.org/259925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/497cca0c8c3661a6c3692630ffda1caabd0da957

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106450 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/15475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/39267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115637 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110358 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16964 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/6700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98649 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112216 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/39267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/39267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/96034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/8707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/39267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6594 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9239 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/6700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/105263 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14861 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/39267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6856 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/10788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebGPU/WebGPU/QuerySet.h:
```diff
@@ -77,6 +80,18 @@ class QuerySet : public WGPUQuerySetImpl, public RefCounted<QuerySet> {
+    // should resolve it from our dummy buffer instead of from where it's supposed to be.
+    //
+    // When rdar://91371495 is fixed, we can delete this indirection, and put the data directly where it's supposed to go.
+    struct Override {
```
Is Override a good name?

litherum replied:
I can rename it 👍

litherum committed:
[WebGPU] Only the first item in PassDescriptor.timestampWrites actually works
https://bugs.webkit.org/show_bug.cgi?id=251779
<rdar://problem/105072144>

Reviewed by Tadeu Zagallo and Dean Jackson.

rdar://91371495 is about how we can't just naively transform descriptor.timestampWrites into computePassDescriptor.sampleBufferAttachments.
Instead, we can resolve all the information to a dummy counter sample buffer, and then internally remember that the data
is in a different place than where it's supposed to be. Later, when we resolve the data, we can resolve it from our dummy
buffer instead of from where it's supposed to be.

When rdar://91371495 is fixed, we can delete this indirection, and put the data directly where it's supposed to go.

This patch causes query-sets.html to work.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateComputePassDescriptor const):
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::resolveQuerySet):
* Source/WebGPU/WebGPU/QuerySet.h:
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::QuerySet::QuerySet):
(WebGPU::QuerySet::destroy):
(WebGPU::QuerySet::setOverride):
(WebGPU::QuerySet::resolveTimestamps const):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::endPass): This is necessary because pass encoders internally retain query sets. If you allocate too many
query sets, trying to create more will fail. So, make sure we don't leave too many query sets hanging around.
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js:

Canonical link: https://commits.webkit.org/259925@main
